### PR TITLE
fixing input validation in segments and delete pit request

### PR DESCRIPTION
**Pull request by bharath-techie**
### Description
Before:
If same PIT ID is sent multiple times as part of delete list or pit segments list, then its processed multiple times.

After this change:
Only the unique PIT IDs are sent from rest layer to transport layer

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Arpit-Bandejiya — APPROVED** [review]

File: server/src/main/java/org/opensearch/action/admin/indices/segments/PitSegmentsRequest.java [hunk 1]
```diff
@@ -87,7 +89,7 @@ public ActionRequestValidationException validate() {
     }
 
     public void fromXContent(XContentParser parser) throws IOException {
-        pitIds.clear();
+        Set<String> pitIds = new HashSet<>();
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/action/admin/indices/segments/PitSegmentsRequest.java:92]: nit- shall we rename the local variable to something else?

**Review by ashking94 — CHANGES_REQUESTED** [review]
can we rename the local pitIds in both the places?

**Review by dhruv16dhr — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -73,6 +73,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Fix 'org.apache.hc.core5.http.ParseException: Invalid protocol version' under JDK 16+ ([#4827](https://github.com/opensearch-project/OpenSearch/pull/4827))
 - Fix compression support for h2c protocol ([#4944](https://github.com/opensearch-project/OpenSearch/pull/4944))
 - Support OpenSSL Provider with default Netty allocator ([#5460](https://github.com/opensearch-project/OpenSearch/pull/5460))
+- Fixing input validation in segments and delete pit request ([#6645](https://github.com/opensearch-project/OpenSearch/pull/6645))
+- Avoid negative memory result in IndicesQueryCache stats calculation ([#6917](https://github.com/opensearch-project/OpenSearch/pull/6917))
```
Comment (Bukhtawar) [inline CHANGELOG.md:77]: Is this intentional?

File: CHANGELOG.md [hunk 3]
```diff
@@ -71,6 +71,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Fix 'org.apache.hc.core5.http.ParseException: Invalid protocol version' under JDK 16+ ([#4827](https://github.com/opensearch-project/OpenSearch/pull/4827))
 - Fix compression support for h2c protocol ([#4944](https://github.com/opensearch-project/OpenSearch/pull/4944))
 - Support OpenSSL Provider with default Netty allocator ([#5460](https://github.com/opensearch-project/OpenSearch/pull/5460))
+- Fixing input validation in segments and delete pit request ([#6645](https://github.com/opensearch-project/OpenSearch/pull/6645))
```
Comment (andrross) [inline CHANGELOG.md:74]: Please move to the `Unreleased 2.x` section and use the imperative mood ("Fix" instead of "Fixing")

**Review by andrross — COMMENTED** [review]
What about all the other code paths in both classes where a list or array is passed in and the values are copied into the ArrayList member? No deduplication is happening in those cases that I can see.

**Comment by andrross** [comment]
> From REST api perspective , handling here seems to be sufficient. Are other code paths affected via user input ?

It just strikes me as odd that we use lists everywhere, but then in two specific cases a set is used internally to dedupe but ultimately a list is still returned by the `getPitIds()`. I would expect one of two approaches:

- Use a java.util.Set internally in each class to enforce uniqueness in the data structure
- Since JSON cannot express a set type, use java.util.Lists so that the full intent of the JSON request is preserved (namely order and duplicates), but then have the consumer of `getPitIds()` explicitly do the de-duplication

I don't doubt that this fixes the bug, but it doesn't feel like a comprehensive fix that would prevent this from happening again.

**Comment by bharath-techie** [comment]
> > From REST api perspective , handling here seems to be sufficient. Are other code paths affected via user input ?
> 
> It just strikes me as odd that we use lists everywhere, but then in two specific cases a set is used internally to dedupe but ultimately a list is still returned by the `getPitIds()`. I would expect one of two approaches:
> 
>     * Use a java.util.Set internally in each class to enforce uniqueness in the data structure
> 
>     * Since JSON cannot express a set type, use java.util.Lists so that the full intent of the JSON request is preserved (namely order and duplicates), but then have the consumer of `getPitIds()` explicitly do the de-duplication
> 
> 
> I don't doubt that this fixes the bug, but it doesn't feel like a comprehensive fix that would prevent this from happening again.

Addressed this - made changes in consumer instead - needed to make changes only in transport layer . security plugin handles duplicates already as we form a map of pitids and indices when decoding PITs.

File: server/src/main/java/org/opensearch/action/admin/indices/segments/TransportPitSegmentsAction.java [hunk 4]
```diff
@@ -94,8 +96,10 @@ public TransportPitSegmentsAction(
      */
     @Override
     protected void doExecute(Task task, PitSegmentsRequest request, ActionListener<IndicesSegmentResponse> listener) {
-        List<String> pitIds = request.getPitIds();
-        if (pitIds.size() == 1 && "_all".equals(pitIds.get(0))) {
+        // remove pit id duplicates from the request
+        Set<String> pitIdSet = new LinkedHashSet<>(request.getPitIds());
+        request.clearAndSetPitIds(new ArrayList<>(pitIdSet));
```
[the remark this answers is not in the document]
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/TransportPitSegmentsAction.java:101]: there are dependencies in security plugin , so i'm trying to find a solution which doesn't change the class contract if possible.

[the remark this answers is not in the document]
Comment (bharath-techie) [reply server/src/main/java/org/opensearch/action/admin/indices/segments/TransportPitSegmentsAction.java]: Security plugin already removes duplicates for PIT ID - need to handle only in transport layer, let me know if the new approach works which doesn't mutate the request.

**Review by andrross — APPROVED** [review]

**Review by Bukhtawar — APPROVED** [review]

**Pull request merged**